File: lua/ffi/c_struct.c
#include <stdio.h>

struct constr_t {
    int a;
    int b;
    struct innerstr {
        int x;
        int y;
    } c;
};

void print_constr_t(struct constr_t t)
{
    printf("a:%d\n", t.a);
    printf("b:%d\n", t.b);
    printf("c.x:%d\n", t.c.x);
    printf("c.y:%d\n", t.c.y);
}

void print_constr_p(struct constr_t *p)
{
    printf("pa:%d\n", p->a);
    printf("pb:%d\n", p->b);
    printf("pc.x:%d\n", p->c.x);
    printf("pc.y:%d\n", p->c.y);
}

File: lua/ffi/c_struct.lua
#! /usr/local/bin/lua
local ffi = require('ffi')
local t = ffi.load("struct", false)

ffi.cdef[[
    struct constr_t {
        int a;
        int b;
        struct innerstr {
            int x;
            int y;
        } c;
    };
    void print_constr_t(struct constr_t t);
    void print_constr_p(struct constr_t *p);
]]

a = ffi.new("struct constr_t", {1, 2, {10, 11}})
t.print_constr_t(a)
t.print_constr_p(a)
